docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+  `cc-shim --token $(token) --uri $(uri)`
+   * The proxy URL, which can be either a UNIX or a TCP socket.
+   * The token for the container process it needs to monitor.
+  The `cc-shim` process will forward the output streams (stderr and stdout) to
+  either `containerd-shim` in the Docker use cases or `conmon` for the Kubernetes
+  deployements.
+4. `cc-runtime` sends an agent `EXECMD` command to start the command in the
+  right container. The command is sent to `cc-proxy` who forwards it to the right
+  agent instance running in the appropriate guest.
+
+Now the `exec`'ed process is running in the virtual machine, sharing the UTS,
+PID, mount and IPC namespaces with the container's init process.
+
+#### [`kill`](https://github.com/clearcontainers/runtime/blob/master/kill.go)
+
+When sending the OCI `kill` command, container runtimes should send a [UNIX signal](https://en.wikipedia.org/wiki/Unix_signal)
+to the container process.
+In the Clear Containers context, this means `cc-runtime` needs to handle 2 resources:
+
+1. It needs to send the signal to the container running on the VM.
+2. It needs to signal the [`shim`](#shim) since we don't want to leave dangling
+  `cc-shim` instances on the host in case the container process terminates inside
+  the VM.
+
+To do so, `cc-runtime` goes through the following steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let
+  it know on which VM the container it is trying to `kill` is running.
+2. `cc-runtime` sends an agent `KILLCONTAINER` command to `kill` the container
+  running on the guest. The command is sent to `cc-proxy` who forwards it to the
+  right agent instance running in the appropriate guest.
+
+After step #2, if the container process terminates in the VM, `cc-proxy` will
+forward the process exit code to the `shim`. The `shim` will then disconnect
+from `cc-proxy` and terminates as well.
+
+One corner case `cc-runtime` also needs to handle is when it receives a signal
+for a container that's not yet running. In that case, `cc-runtime` simply stops
+`cc-shim` for the `SIGTERM` or `SIGKILL` signals and does nothing for all other
+signals.
+
+#### [`delete`](https://github.com/clearcontainers/runtime/blob/master/delete.go)
+
+`delete` is about deleting all resources held by a stopped/killed container.
+Running containers can not be deleted unless the OCI runtime is explictly being
+asked to. In that case it will first `kill` the container and only then `delete`
+it.
+
+The `delete` code path differs significantly between having to delete one container
+inside a pod and having to delete an entire pod. In the former case, `cc-runtime`
+will only send a `SIGKILL` signal to the container process while the latter case is
+about completely deleting the whole thing: the pod, its containers, all `cc-shim`
+instances and the virtual machine itself.
+Also, the latter case is only supported by Kubernetes whereas Docker always asks
+for the former one.
```

Comment:
This makes it sound slightly like Docker will never ask for the cc-shim and VM to be deleted - I don't know if we nee to clarify that a little here.

Reply:
@grahamwhaley @sameo @jodh-intel - I agree, I don't like how it is worded as is.  I was moving towards:
```
328 The `delete` code path differs significantly between having to delete one container
329 inside a pod (as is typical in Docker) and having to delete an entire pod (which
330 is unique to Kubernetes). In the former case, `cc-runtime` will only send a   
331 `SIGKILL` signal to the container process.  In the latter case, the whole thing 
332 is deleted: the pod, its containers, all `cc-shim` instances and the virtual  
333 machine itself.    
```
--but this still doesn't really address your point.  Suggestions?